Refactor controller initialization (round #3)

mfojtik opened the pull request:
Follow up for https://github.com/openshift/origin/pull/14293 with more controllers.

This time:

* image trigger controller
* service serving cert controller
* image import controller

mfojtik commented:
@deads2k https://github.com/openshift/origin/pull/14293 is going to have green tests, lets continue refactoring here. i'm going to move image controllers. @soltysh FYI you don't have to do this.

mfojtik commented:
[test]

mfojtik commented:
[test]

mfojtik commented:
@deads2k ready for the round 3 ?

mfojtik commented:
@soltysh for the image controllers

mfojtik commented:
@smarterclayton for the image trigger controller

mfojtik commented on pkg/cmd/server/origin/controller/image.go:
```diff
@@ -0,0 +1,141 @@
+	return true, nil
+}
+
+// TODO: remove when generated informers exist
```
@deads2k @smarterclayton do they exists now? i saw some image stream generated informers already ;)

soltysh replied:
They exists, but are not wired, yet (or not everywhere) iirc. This should be removed when we entirely switch to the generated informers. The question is do we want to make it happen for 3.6?

mfojtik replied:
no, that is 3.7

mfojtik commented on pkg/cmd/server/origin/controller/image.go:
```diff
@@ -139,3 +141,39 @@ func (u podSpecUpdater) Update(obj runtime.Object) error {
+			Resync: time.Duration(c.ScheduledImageImportMinimumIntervalSeconds) * time.Second,
+
+			Enabled:                  !c.DisableScheduledImport,
+			DefaultBucketSize:        4, // TODO: Make this configurable?
```
@soltysh do we have issue for this? do we care?

soltysh replied:
IIRC, you've added that TODO, I don't think we care. Drop the todo.

soltysh commented on pkg/cmd/server/bootstrappolicy/infra_sa_policy.go:
```diff
@@ -13,28 +13,26 @@ import (
-	InfraDeploymentConfigControllerServiceAccountName          = "deploymentconfig-controller"
-	InfraDeploymentTriggerControllerServiceAccountName         = "deployment-trigger-controller"
-	InfraDeployerControllerServiceAccountName                  = "deployer-controller"
+	// The controllers below were convered to new controller initialization and use RBAC
```
Typo: conver**t**ed

soltysh commented on test/integration/buildcontroller_test.go:
```diff
@@ -165,7 +165,10 @@ func setupBuildControllerTest(counts controllerCount, t *testing.T) (*client.Cli
 	}
 	for i := 0; i < counts.ImageChangeControllers; i++ {
-		openshiftConfig.RunImageTriggerController()
+		_, err := openshiftControllerInitializers["imagetrigger"](openshiftControllerContext)
```
I'm guessing this should be `image-trigger` since in the 3rd commit you're changing that.

mfojtik replied:
yes, good catch... altough i'm not sure how this will not panic ;-)

deads2k commented on pkg/cmd/server/origin/controller.go:
```diff
@@ -115,8 +118,36 @@ func (c *MasterConfig) NewOpenshiftControllerInitializers() (map[string]controll
+		HasStatefulSetsEnabled: c.Options.DisabledFeatures.Has("triggers.image.openshift.io/statefulsets"),
+		HasCronJobsEnabled:     c.Options.DisabledFeatures.Has("triggers.image.openshift.io/cronjobs"),
+	}
+	ret["image-trigger"] = imageTrigger.RunController
```
openshift.io/*

deads2k commented on pkg/cmd/server/origin/controller.go:
```diff
@@ -115,8 +118,36 @@ func (c *MasterConfig) NewOpenshiftControllerInitializers() (map[string]controll
+		DisableScheduledImport:                     c.Options.ImagePolicyConfig.DisableScheduledImport,
+		ScheduledImageImportMinimumIntervalSeconds: c.Options.ImagePolicyConfig.ScheduledImageImportMinimumIntervalSeconds,
+	}
+	ret["image-import"] = imageImport.RunController
```
openshift.io/*

deads2k commented on pkg/cmd/server/origin/controller.go:
```diff
@@ -115,8 +118,36 @@ func (c *MasterConfig) NewOpenshiftControllerInitializers() (map[string]controll
+			CertFile: c.Options.ControllerConfig.ServiceServingCert.Signer.CertFile,
+			KeyFile:  c.Options.ControllerConfig.ServiceServingCert.Signer.KeyFile,
+		}
+		ret["service-serving-cert"] = serviceServingCert.RunController
```
openshift.io/*

deads2k commented:
I'd like to see https://github.com/openshift/origin/pull/14333 merge first.  Can you rebase on top of it?

deads2k commented on pkg/cmd/server/origin/controller/image.go:
```diff
@@ -0,0 +1,179 @@
+	}
+	if !c.HasDeploymentsEnabled {
+		sources = append(sources, imagetriggercontroller.TriggerSource{
+			Resource:  schema.GroupResource{Group: "extensions", Resource: "deployments"},
```
open an issue to update this to apps or we're going to have a problem.

mfojtik replied:
https://github.com/openshift/origin/issues/14543

deads2k commented on pkg/cmd/server/origin/controller/image.go:
```diff
@@ -0,0 +1,179 @@
+	case *kappsv1beta1.StatefulSet:
+		_, err := u.kclient.Apps().StatefulSets(t.Namespace).Update(t)
+		return err
+	case *kapiv1.Pod:
```
we're touching pods for some reason?

mfojtik replied:
@smarterclayton ^

deads2k commented on pkg/cmd/server/origin/controller/image.go:
```diff
@@ -0,0 +1,179 @@
+
+	go controller.Run(5, ctx.Stop)
+	if c.DisableScheduledImport {
+		return true, nil
```
move to the beginning of the function.  `return false, nil`

deads2k commented on pkg/cmd/server/origin/controller/service.go:
```diff
@@ -0,0 +1,45 @@
+	KeyFile  string
+}
+
+func (c *ServiceServingCertsControllerOptions) RunController(ctx ControllerContext) (bool, error) {
```
if there's no cert and no key, `return false, nil`

deads2k commented on pkg/cmd/server/origin/controller/service.go:
```diff
@@ -0,0 +1,45 @@
+func (c *ServiceServingCertsControllerOptions) RunController(ctx ControllerContext) (bool, error) {
+	ca, err := crypto.GetCA(c.CertFile, c.KeyFile, "")
+	if err != nil {
+		return false, fmt.Errorf("service serving cert controller: %v", err)
```
`return true, fmt.Errorf...`

deads2k commented on pkg/cmd/server/origin/controller.go:
```diff
@@ -115,8 +118,36 @@ func (c *MasterConfig) NewOpenshiftControllerInitializers() (map[string]controll
 	ret["templateinstance"] = templateInstance.RunController
 
+	if c.Options.ControllerConfig.ServiceServingCert.Signer != nil {
+		serviceServingCert := controller.ServiceServingCertsControllerOptions{
```
I think I prefer an unconditional add here with the skip happening in the init function.  That way the list of these things remains consistent.

mfojtik replied:
@deads2k so pass the Signer via opts?

mfojtik commented:
[merge] lgtemd by @deads2k per IRC

deads2k commented on pkg/cmd/server/origin/controller/image.go:
```diff
@@ -154,6 +154,11 @@ func (c *ImageImportControllerOptions) RunController(ctx ControllerContext) (boo
 		ctx.ClientBuilder.DeprecatedOpenshiftClientOrDie(bootstrappolicy.InfraImageImportControllerServiceAccountName),
 		ctx.DeprecatedOpenshiftInformers.ImageStreams(),
 	)
+	go controller.Run(5, ctx.Stop)
```
Oh, two different controllers?

mfojtik replied:
yeah, depending on each-other

deads2k commented:
One comment and lgtm, squash the comments commit.

mfojtik commented:
@deads2k or @smarterclayton can you please review the start_master and the image controller file? thats where the most conflict were, I hope i fixed them properly and don't miss anything.

deads2k commented:
> @deads2k or @smarterclayton can you please review the start_master and the image controller file? thats where the most conflict were, I hope i fixed them properly and don't miss anything.

still lgtm

mfojtik commented:
@enj

mfojtik commented:
[merge]

mfojtik commented:
@deads2k @smarterclayton now that i fixed the build integration test to incorporate the informers and rebased this, i'm back in position 10... i'm trying to merge this for ~8 days now and it is becoming hell to rebase this and get back to the beggining on the queue every morning...

can we green button merge this? i'm willing to rebase this once again after generated auth informers go in, but if we want this in 3.6, it would be better to green button merge it.

deads2k commented:
> @deads2k @smarterclayton now that i fixed the build integration test to incorporate the informers and rebased this, i'm back in position 10... i'm trying to merge this for ~8 days now and it is becoming hell to rebase this and get back to the beggining on the queue every morning...
> 
> can we green button merge this? i'm willing to rebase this once again after generated auth informers go in, but if we want this in 3.6, it would be better to green button merge it.

@smarterclayton @eparis @pweil-   this has happened to multiple pulls over the past two sprints and is symptomatic of our larger queue problems.  While the green button provides us with some degree of control, manual queue management like this is unsustainable in the long run and inherently favoritist.

mfojtik commented:
@jupierce ^

mfojtik commented:
@smarterclayton @eparis @jupierce another rebase, another queue bump: You are in the build queue at position: 18 :-(

enj commented:
[severity: blocker]

mfojtik commented:
@enj ❤️

mfojtik commented:
broken by https://github.com/openshift/origin/pull/14289 ...

deads2k commented:
> [severity:blocker]

Agree

smarterclayton commented:
Compile error
